software/utils/debug/debug_cli.py

```diff
@@ -0,0 +1,645 @@
+   Its a string made up of the following characters:
+     +  current address
+     x  16-bit integer displayed in hex
+     i  16-bit signed integer displayed in decimal
+     u  16-bit unsigned integer displayed in decimal
+     X  32-bit integer displayed in hex
+     I  32-bit signed integer displayed in decimal
+     U  32-bit unsigned integer displayed in decimal
+     f  32-bit float
+     e  32-bit float in exponential format
+     c  Single byte displayed as an ASCII character
+     b  Single byte displayed in hex
+
+   The data is extracted from what's returned and formatted as described in the string.
+   If there's more data left over at the end of the string, a new line starts and the
+   string starts over.
+
+   The default formatting string if none is supplied is +XXXX
+   i.e. Data is displayed as a series of 4 32-bit hex values / line
+"""
+        param = shlex.split(line)
+        if len(param) < 1:
+            print("Please specify the address at which to peek at a minimum")
+            return
+        ct = 1
+        fmt = "+XXXX"
+        file_name = None
+        if len(param) > 3:
+            file_name = param[3]
+        if len(param) > 2:
+            fmt = param[2]
+        if len(param) > 1:
+            ct = int(param[1], 0)
+        address = int(param[0], 0)
+        self.interface.peek(address, ct, fmt, file_name)
+
+    def do_poke(self, line):
+        """Write data to a memory address
+
+ex: poke [type] <address> <data>
+
+   type    - Optional type, can be byte, short, long or float
+             determines how the data will be interpreted.
+
+   address - Address at which to write data
+
+   data    - One or more data items to write.
+"""
+        param = shlex.split(line)
+        if len(param) < 2:
+            print("Please pass the address and at least one value to write")
+            return
+
+        poke_type = "byte"
+        if param[0] in ["long", "short", "float"]:
+            poke_type = param[0]
+            param = param[1:]
+
+        address = param[0]
+
+        if poke_type == "float":
+            data = [float(x) for x in param[1:]]
+        else:
+            data = [int(x, 0) for x in param[1:]]
+        self.interface.poke(address, data, poke_type)
+
+    def do_EOF(self, line):
+        return True
+
+    def do_exit(self, line):
+        return True
+
+    def do_get(self, line):
+        cl = line.split()
+        if len(cl) < 1:
+            print("Please give the variable name to read")
+            return
+
+        if cl[0] == "all":
+            all_vars = self.interface.variables_get_all()
+            for name, val in all_vars.items():
+                print(f"{name:25} = {val}")
+            return
+
+        if len(cl) > 1:
+            fmt = cl[1]
+        else:
+            fmt = None
+
+        print(self.interface.variable_get(cl[0], fmt=fmt))
+
+    def complete_get(self, text, line, begidx, endidx):
```

Comment:
yes, we can write these for `trace` and `test` as well, but it felt a little tedious, because there are sub-commands with different parameters and I haven't wrapped my head around how to do that. I don't think you want to initially generate every permutation? And not sure what the index parameters are for. Seems nontrivial. Maybe new ticket for this?